Improve search speed

sascha-karnatz opened the pull request:
Add a `tsvector` column to store the tsvector - content in the database. This speeds up the search speed on large table, because these value isn't calculated on fly anymore. A dictionary configuration to allow to setup the used PostgreSQL dictionary.

sascha-karnatz committed:
Improve the search speed

Add tsvactor_column to store the tsvector - content in the database. This speeds up the search speed on large table, because these value isn't calculated on fly anymore. A dictionary configuration to allow to setup the used PostgreSQL dictionary.

tvdeyen commented on db/migrate/20231218165617_add_searchable_content.rb:
```diff
@@ -0,0 +1,16 @@
+    unless column_exists? :pg_search_documents, :searchable_content
+      execute %|ALTER TABLE pg_search_documents
+        ADD COLUMN searchable_content tsvector GENERATED ALWAYS AS (
+          to_tsvector('#{Alchemy::PgSearch.config[:dictionary]}', coalesce(content, ''))
```
```suggestion
          to_tsvector('#{Alchemy::PgSearch.config.fetch(:dictionary, "simple")}', coalesce(content, ''))
```

tvdeyen commented on spec/dummy/db/migrate/20231218165617_add_searchable_content.rb:
```diff
@@ -0,0 +1,16 @@
+class AddSearchableContent < ActiveRecord::Migration[6.0]
+  def up
+    unless column_exists? :pg_search_documents, :searchable_content
+      execute %|ALTER TABLE pg_search_documents
+        ADD COLUMN searchable_content tsvector GENERATED ALWAYS AS (
+          to_tsvector('#{Alchemy::PgSearch.config[:dictionary]}', coalesce(content, ''))
+        ) STORED;|
```
Can we try to use Rails 7 feature of adding a generated column?

```suggestion
    add_column :pg_search_documents, :searchable_content, :virtual, type: :tsvector,
      as: "to_tsvector('#{Alchemy::PgSearch.config.fetch(:dictionary, "simple")}', coalesce(content, ''))",
      stored: true,
      if_not_exists: true
```

tvdeyen commented on spec/dummy/db/migrate/20231218165617_add_searchable_content.rb:
```diff
@@ -0,0 +1,16 @@
+class AddSearchableContent < ActiveRecord::Migration[6.0]
+  def up
```
We should be able to use `change`

tvdeyen reviewed (commented):
Nice! I think we can make use of Rails 7 virtual column support and need to update the postgres version in the CI to 12

sascha-karnatz committed:
Require at least PostgreSQL v12

The tsvector - column is only available in PostgreSQL v12 or higher. Add a requirement section in the README and adjusted the CI workflow.

tvdeyen reviewed (approved):
This is amazing
